compiler/rustc_hir_analysis/src/check/always_applicable.rs

```diff
@@ -0,0 +1,178 @@
+use rustc_data_structures::fx::FxHashSet;
+use rustc_errors::{struct_span_err, ErrorGuaranteed};
+use rustc_infer::infer::outlives::env::OutlivesEnvironment;
+use rustc_infer::infer::{RegionResolutionError, TyCtxtInferExt};
+use rustc_middle::ty::subst::SubstsRef;
+use rustc_middle::ty::util::CheckRegions;
+use rustc_middle::ty::{self, TyCtxt};
+use rustc_trait_selection::traits::{self, ObligationCtxt};
+
+use crate::hir::def_id::DefId;
+
+/// This function confirms that the trait implementation identified by
+/// `trait_impl_def_id` is not any more specialized than the type it is
+/// attached to (Issue #8142).
+///
+/// This means:
+///
+/// 1. The self type must be nominal (this is already checked during
+///    coherence),
+///
+/// 2. The generic region/type parameters of the impl's self type must
+///    all be parameters of the Trait impl itself (i.e., no
+///    specialization like `impl Trait for Foo<i32>`), and,
+///
+/// 3. Any bounds on the generic parameters must be reflected in the
+///    struct/enum definition for the nominal type itself (i.e.
+///    cannot do `struct S<T>; impl<T:Clone> Trait for S<T> { ... }`).
+///
+pub fn check_trait_impl(tcx: TyCtxt<'_>, trait_impl_def_id: DefId) -> Result<(), ErrorGuaranteed> {
+    let self_type = tcx.type_of(trait_impl_def_id).subst_identity();
+    match self_type.kind() {
+        ty::Adt(adt_def, adt_to_impl_substs) => {
+            ensure_trait_params_and_item_params_correspond(
+                tcx,
+                trait_impl_def_id,
+                adt_def.did(),
+                adt_to_impl_substs,
+            )?;
+
+            ensure_trait_predicates_are_implied_by_item_defn(
+                tcx,
+                trait_impl_def_id,
+                adt_def.did(),
+                adt_to_impl_substs,
+            )
+        }
+
+        _ => Ok(()),
```

Comment:
What about negative impls like `impl !Trait for &T`? Drop impls don't care about these because we separately enforce they must be implemented for only ADTs, but negative impls don't have this requirement, do they?

Reply:
I've added coverage for this and other relevant cases.

Reply:
Sorry, test coverage? I don't see it :thinking: